src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt

```diff
@@ -13,124 +13,123 @@ import com.intellij.openapiext.isUnitTestMode
 
-                val validParsedEnd = matchResult.range.last + 1
-                if (validParsedEnd != inside.value.length) {
-                    // BACKCOMPAT: 2019.3
-                    @Suppress("DEPRECATION")
-                    holder.createErrorAnnotation(
-                        (validParsedEnd until inside.value.length).toSourceRange(),
-                        "Invalid format string"
-                    )
-                } else {
-                    holder.highlightArgument(matchResult.groups["id"]?.range)
-                    holder.highlightArgument(matchResult.groups["width"]?.range)
-                    holder.highlightArgument(matchResult.groups["precision"]?.range)
-                    holder.highlightArgument(matchResult.groups["type"]?.range, RsColor.FUNCTION)
-                    //TODO: argument type/count/other checks
-                }
-            }
+        // skip advanced checks and highlighting if there are syntax errors
+        if (errors.isNotEmpty()) {
+            return
+        }
+
+        highlightParametersInside(parseCtx, holder)
+
+        if (!isUnitTestMode && element.project.macroExpansionManager.macroExpansionMode !is MacroExpansionMode.New) return
+
+        val parameters = buildParameters(parseCtx)
+        val arguments = macroArgs
+            .drop(macroPos + 1)
+            .toList()
+        val ctx = FormatContext(parameters, arguments, formatMacro)
+
+        val annotations = checkParameters(ctx).toMutableList()
+        annotations += checkArguments(ctx)
+
+        for (annotation in annotations) {
+            holder.createErrorAnnotation(annotation.range, annotation.error)
         }
     }
+}
 
-    private fun macroToFormatPos(macro: String): Int = when (macro) {
-        "println" -> 0
-        "print" -> 0
-        "eprintln" -> 0
-        "eprint" -> 0
-        "format" -> 0
-        "format_args" -> 0
-        "format_args_nl" -> 0
-        "write" -> 1
-        "writeln" -> 1
-        else -> -1
+private data class ParameterMatchInfo(val range: TextRange, val match: MatchResult, val type: String)
+
+private sealed class FormatParameter(val matchInfo: ParameterMatchInfo) {
+    override fun toString(): String {
+        return this.matchInfo.match.value
+    }
+
+    val range: TextRange = matchInfo.range
+
+    fun requiredTrait(knownItems: KnownItems): RsTraitItem? {
+        return when (this.matchInfo.type) {
+            "" -> knownItems.Display
+            "?", "x?", "X?" -> knownItems.Debug
+            "o" -> knownItems.Octal
+            "x" -> knownItems.LowerHex
+            "X" -> knownItems.UpperHex
+            "p" -> knownItems.Pointer
+            "b" -> knownItems.Binary
+            "e" -> knownItems.LowerExp
+            "E" -> knownItems.UpperExp
+            else -> return null
+        }
     }
+}
+private class NamedParameter(matchInfo: ParameterMatchInfo, val name: String) : FormatParameter(matchInfo)
```

Comment:
We usually put sealed class children into the parent sealed class. I.e. `FormatParameter.Named` and `FormatParameter.Positional`